velox/exec/Spill.h

```diff
@@ -0,0 +1,340 @@
+    }
+    return {&decoded_, rowVector_.get(), index_++};
+  }
+
+  virtual uint64_t size() const {
+    return 0;
+  }
+
+ protected:
+  virtual void nextBatch() = 0;
+
+  std::shared_ptr<const RowType> type_;
+  memory::MemoryPool& pool_;
+  RowVectorPtr rowVector_;
+  vector_size_t index_ = 0;
+  vector_size_t numRows_ = 0;
+  std::vector<std::unique_ptr<DecodedVector>> decoded_;
+  int32_t ordinal_;
+  static std::atomic<int32_t> ordinalCounter_;
+};
+
+// Represents a spill file that is first in write mode and then
+// turns into an source of VectorRow. Owns a file system file that
+// contains the spilled data and is live for the duration of 'this'.
+class SpillFile : public SpillStream {
+ public:
+  SpillFile(TypePtr type, const std::string& path, memory::MemoryPool& pool)
+      : SpillStream(type, pool), path_(path) {}
+
+  ~SpillFile();
+
+  // Returns a file for writing spilled data.
+  WriteFile& output();
+
+  // Finishes writing and flushes any unwritten data.
+  void finishWrite() {
+    VELOX_CHECK(output_);
+    size_ = output_->size();
+    output_ = nullptr;
+  }
+
+  // Prepares 'this' for reading. Positions the read at the first row of
+  // content.
+  void startRead();
+
+  // Returns the file size in bytes.
+  uint64_t size() const override {
+    if (output_) {
+      return output_->size();
+    }
+    return size_;
+  }
+
+  // True if no more VectorRows can be read from 'this'.
+  bool atEnd() const override {
+    return index_ >= numRows_ && stream_->atEnd();
+  }
+
+  // Sets 'result' to refer to the next row of content of 'this'.
+  void read(RowVector& result);
+
+ protected:
+  void nextBatch() override {
+    VectorStreamGroup::read(stream_.get(), &pool_, type_, &rowVector_);
+    index_ = 0;
+    numRows_ = rowVector_->size();
+  }
+
+ private:
+  std::string path_;
+  uint64_t size_ = 0;
+  std::unique_ptr<WriteFile> output_;
+  std::unique_ptr<SpillInput> stream_;
+};
+
+// Describes a bit field inside a 64 bit hash number for use in
+// partitioning data over multiple sets of spill files.
+struct HashBitField {
+  // Low bit number of hash number bit field
+  uint8_t begin;
+  // Bit number of first bit above the hash number bit field.
+  uint8_t end;
+};
+
+// Sequence of files for one partition of the spilled data. If data is
+// sorted, each file is sorted. The globally sorted order is produced
+// by merging the constituent files.
+class FileList {
+ public:
+  // Constructs a set of spill files. 'type' a RowType describing the content.
+  // 'path' is a file path prefix.  FileList( 'targetBatchSize is the target
+  // size of a single RowVector in rows. 'targetFileSize' is the target byte
+  // size of a single file in the file set. 'pool' and 'mappedMemory' are used
+  // for buffering and constructing the result data read from 'this'.
+  FileList(
+      TypePtr type,
+      const std::string path,
+      uint64_t targetBatchSize,
+      uint64_t targetFileSize,
+      memory::MemoryPool& pool,
```

Comment:
It would be nice to use either references or raw pointers for MemoryPool and MappedMemory, but not both.